core/src/main/java/cz/o2/proxima/util/SerializableScopedValue.java

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.util;
+
+import com.google.common.base.MoreObjects;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.ObjectStreamException;
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A value that holds a serializable value and scopes it value to given context value.
+ *
+ * @param <C> context type parameter
+ * @param <V> type parameter
+ */
+public final class SerializableScopedValue<C, V extends Serializable> implements Serializable {
+
+  private static final Map<Integer, Map<Object, Object>> MAP = new ConcurrentHashMap<>();
+
+  private final int identityHashCode = System.identityHashCode(this);
```

Comment:
Two instances will result in two different objects, because `identityHashCode` of `this` will be different. `UUID` is possible, but because the "driver" JVM should always be single JVM, that should be semantically equivalent.